### PR TITLE
Handleactivitymodule

### PR DESCRIPTION
**Pull request by vishaltambe70**
activity /activity assignee module.

### CODE REVIEW

File: crm-plugin/config/routes/activity.json [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      "path": "/",
+      "handler": "activity.index",
+      "config": {
+        "policies": []
+      }
+    },
+    {
+      "method": "GET",
+      "path": "/activities/:id",
+      "handler": "activity.find",
```
Comment (tanviGanesh) [inline crm-plugin/config/routes/activity.json]: this action should be for findOne method in controller. refer contact.js

File: crm-plugin/controllers/activity.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  },
+
+  find: async (ctx) => {
+    let activity;
+    if (ctx.query._q) {
+      console.log("sdfhjdsfhjfdshj", ctx.query._q);
+      activity = await strapi.query("activity", "crm-plugin").search(ctx.query);
+    } else {
+      activity = await strapi.query("activity", "crm-plugin").find(ctx.query);
+    }
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activity.js:60]: error handling not done. refer contact.js controller

File: crm-plugin/controllers/activity.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    // Send 200 `ok`
+    ctx.send({
+      message: "ok",
+    });
+  },
+
+  find: async (ctx) => {
+    let activity;
+    if (ctx.query._q) {
+      console.log("sdfhjdsfhjfdshj", ctx.query._q);
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activity.js]: remove console statements from file please

File: crm-plugin/controllers/activity.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            .query("activityassignee", "crm-plugin")
+            .create(activityDetail);
+        }
+        return sanitizeEntity(activity, {
+          model: strapi.plugins["crm-plugin"].models["activity"],
+        });
+      }
+    } catch (error) {
+      console.error(error);
+      return { error: error.message };
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activity.js]: handle error using badRequest

File: crm-plugin/controllers/activity.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    let activityassignee;
+    let activity;
+    try {
+      if (ctx.params.id) {
+        const { id } = ctx.params;
+        activity = await strapi
+          .query("activity", "crm-plugin")
+          .update({ id }, ctx.request.body);
+          if(ctx.request.body.contact){
+            let activityDetail = {
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activity.js]: order wrong. contact and activityassignee has to be updated first and then activity

File: crm-plugin/controllers/activity.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        } else {
+          activity = await strapi
+            .query("activity", "crm-plugin")
+            .create(ctx.request.body);
+          if (ctx.request.body.contact) {
+            let activityDetail = {
+              activity: activity.id,
+              contact: ctx.request.body.contact,
+            };
+            activityassignee = await strapi
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activity.js]: order wrong. contact and activityassignee has to be updated first and then activity

File: crm-plugin/controllers/activity.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+              .query("activityassignee", "crm-plugin")
+              .update({ activity: ctx.params.id }, activityDetail);
+          }
+          return sanitizeEntity(activity, {
+            model: strapi.plugins["crm-plugin"].models["activity"],
+          });
+        } else {
+          activity = await strapi
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activity.js]: call strapi.plugins["crm-plugin"].services.utils.checkParams method refer contact.js controller

File: crm-plugin/controllers/activityassignee.js [hunk 8]
```diff
@@ -0,0 +1,7 @@
+"use strict";
+/* jshint node:true */
+
+var Base = require("./common/base");
+
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activityassignee.js:5]: handle required params

File: crm-plugin/testcases/activity.test.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+  /* this method calls all testcases for empty params check */
+  this.emptyTestcase = async (method) => {
+    let methodParams = {};
+    //this switch case needs to handle all methods of state controller
+    switch (method) {
+      case "find":
+        methodParams = {
+          originalUrl: "/crm-plugin/activities",
+          query: {},
```
Comment (tanviGanesh) [inline crm-plugin/testcases/activity.test.js]: handle badrequest method

Comment (tanviGanesh) [reply crm-plugin/testcases/activity.test.js]: also handle other methods for required and correct params

File: crm-plugin/controllers/activity.js (hunk 3 above)
Comment (vishaltambe70) [reply crm-plugin/controllers/activity.js]: done

File: crm-plugin/controllers/activity.js (hunk 4 above)
Comment (vishaltambe70) [reply crm-plugin/controllers/activity.js]: done

File: crm-plugin/controllers/activity.js (hunk 5 above)
Comment (vishaltambe70) [reply crm-plugin/controllers/activity.js]: done

File: crm-plugin/controllers/activity.js (hunk 6 above)
Comment (vishaltambe70) [reply crm-plugin/controllers/activity.js]: done

File: crm-plugin/controllers/activity.js (hunk 5 above)
Comment (vishaltambe70) [reply crm-plugin/controllers/activity.js:83]: done

File: crm-plugin/controllers/activity.js (hunk 7 above)
Comment (vishaltambe70) [reply crm-plugin/controllers/activity.js]: done

File: crm-plugin/testcases/activity.test.js (hunk 9 above)
Comment (vishaltambe70) [reply crm-plugin/testcases/activity.test.js]: done

File: crm-plugin/controllers/activity.js (hunk 5 above)
Comment (tanviGanesh) [reply crm-plugin/controllers/activity.js]: order not changed....activity table updated first currently please do properly

File: crm-plugin/controllers/activity.js (hunk 7 above)
Comment (tanviGanesh) [reply crm-plugin/controllers/activity.js]: not done

File: crm-plugin/controllers/activity.js (hunk 6 above)
Comment (tanviGanesh) [reply crm-plugin/controllers/activity.js]: order not changed

File: crm-plugin/controllers/activityassignee.js (hunk 8 above)
Comment (vishaltambe70) [reply crm-plugin/controllers/activityassignee.js:5]: done

File: crm-plugin/controllers/activity.js [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        if (ctx.request.body.contact) {
+          let activityDetail = {
+            activity: ctx.params.id,
+            contact: ctx.request.body.contact,
+          };
+          activityassignee = await strapi
+            .query("activityassignee", "crm-plugin")
+            .update({ activity: ctx.params.id }, activityDetail);
+        }
+        const id = activityassignee.id;
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activity.js]: wrong..how can you update activity table with activityassignee???

Comment (tanviGanesh) [reply crm-plugin/controllers/activity.js]: it should be id of activity table na ??? havent you tested this?  its completely wrong code

File: crm-plugin/controllers/activity.js [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        activity = await strapi
+          .query("activity", "crm-plugin")
+          .update(id, ctx.request.body);
+        return sanitizeEntity(activity, {
+          model: strapi.plugins["crm-plugin"].models["activity"],
+        });
+      } else {
+        const demoParams = {
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activity.js]: why this is declared?? remove code if not used

File: crm-plugin/controllers/activity.js [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        };
+        const reqVal = ["title"];
+        const result = strapi.plugins["crm-plugin"].services.utils.checkParams(
+          ctx.request.body,
+          reqVal
+        );
+        if (result.error == true) {
+          return ctx.badRequest(null, result.message);
+        }
+        activity = await strapi
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activity.js:139]: order needs to be changed here. first entry in activity assignee table then activity table

File: crm-plugin/controllers/activity.js [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    try {
+      let ids = ctx.params;
+      const activityass = await strapi
+        .query("activityassignee", "crm-plugin")
+        .delete({ activity: ctx.params.id });
+      const activity = await strapi
+        .query("activity", "crm-plugin")
+        .delete(ctx.params);
+      return sanitizeEntity(activity, {
+        model: strapi.plugins["crm-plugin"].models["activityassignee"],
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activity.js]: activity model should be returned always... this is completely wrong

File: crm-plugin/testcases/activity.test.js [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+          badRequest: (error, message) => {
+            return { error: message };
+          },
+        };
+        break;
+      case "create":
+        methodParams = {
```
Comment (tanviGanesh) [inline crm-plugin/testcases/activity.test.js]: findOne method not handled

File: crm-plugin/testcases/activity.test.js [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+    switch (method) {
+      case "find":
+        methodParams = {
+          query: {},
+          badRequest: (error, message) => {
+            return { error: message };
+          },
+        };
+        console.log("Required params test not applicable for find method");
+        console.log("-------------");
```
Comment (tanviGanesh) [inline crm-plugin/testcases/activity.test.js]: need to return here check state.test.js file

File: crm-plugin/testcases/activity.test.js [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+          },
+        };
+        break;
+      case "delete":
+        methodParams = {
+          params: {},
+          badRequest: (error, message) => {
+            return { error: message };
+          },
+        };
```
Comment (tanviGanesh) [inline crm-plugin/testcases/activity.test.js]: need to return here too. wrongly done

File: crm-plugin/testcases/activity.test.js [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+    //this switch case needs to handle all methods of contact controller
+    switch (method) {
+      case "find":
+        methodParams = {
+          query: {},
+          badRequest: (error, message) => {
+            return { error: message };
+          },
+        };
+        console.log("Correct params test not applicable for find method");
```
Comment (tanviGanesh) [inline crm-plugin/testcases/activity.test.js]: correct params method is proper

**Review by tanviGanesh — CHANGES_REQUESTED** [review]

File: crm-plugin/controllers/activity.js [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+
+    try {
+      if (ctx.params.id) {
+        if (ctx.request.body.contact) {
+          let activityDetail = {
+            activity: ctx.params.id,
+            contact: ctx.request.body.contact,
+          };
+          activityassignee = await strapi
+            .query("activityassignee", "crm-plugin")
```
Comment (tanviGanesh) [inline crm-plugin/controllers/activity.js]: issue here. what if activity assignee is passed to update api first time. this will fail.

**Review by tanviGanesh — CHANGES_REQUESTED** [review]

**Comment by tanviGanesh** [comment]
Unable to check the feature as there is runtime error when tried to run strapi server on this branch. 
image

**Pull request merged**